src/cost_estimates/fee_medians.rs

```diff
@@ -0,0 +1,348 @@
+};
+use serde_json::Value as JsonValue;
+
+use chainstate::stacks::TransactionPayload;
+use util::db::sqlite_open;
+use util::db::tx_begin_immediate_sqlite;
+use util::db::u64_to_sql;
+
+use vm::costs::ExecutionCost;
+
+use chainstate::stacks::db::StacksEpochReceipt;
+use chainstate::stacks::events::TransactionOrigin;
+
+use crate::util::db::sql_pragma;
+use crate::util::db::table_exists;
+
+use super::metrics::CostMetric;
+use super::FeeRateEstimate;
+use super::{EstimatorError, FeeEstimator};
+
+use super::metrics::PROPORTION_RESOLUTION;
+use cost_estimates::StacksTransactionReceipt;
+
+const CREATE_TABLE: &'static str = "
+CREATE TABLE median_fee_estimator (
+    measure_key INTEGER PRIMARY KEY AUTOINCREMENT,
+    high NUMBER NOT NULL,
+    middle NUMBER NOT NULL,
+    low NUMBER NOT NULL
+)";
+
+/// FeeRateEstimator with the following properties:
+///
+/// 1) We use a "weighted" percentile approach for calculating the percentile values. Described
+///    below, larger transactions contribute more to the ranking than small transactions.
+/// 2) Use "windowed" decay instead of exponential decay. This allows outliers to be forgotten
+///    faster, and so reduces the influence of outliers.
+/// 3) "Pad" the block, so that any unused spaces is considered to have an associated fee rate of
+///    1f, the minimum. Ignoring the amount of empty space leads to over-estimates because it
+///    ignores the fact that there was still space in the block.
+pub struct WeightedMedianFeeRateEstimator<M: CostMetric> {
+    db: Connection,
+    /// We only look back `window_size` fee rates when averaging past estimates.
+    window_size: u32,
+    /// The weight of a "full block" in abstract scalar cost units. This is the weight of
+    /// a block that is filled on each dimension.
+    full_block_weight: u64,
+    /// Use this cost metric in fee rate calculations.
+    metric: M,
+}
+
+/// Convenience struct for passing around this pair.
+#[derive(Debug)]
+struct FeeRateAndWeight {
+    pub fee_rate: f64,
+    pub weight: u64,
+}
+
+impl<M: CostMetric> WeightedMedianFeeRateEstimator<M> {
+    /// Open a fee rate estimator at the given db path. Creates if not existent.
+    pub fn open(p: &Path, metric: M, window_size: u32) -> Result<Self, SqliteError> {
+        let mut db = sqlite_open(
+            p,
+            rusqlite::OpenFlags::SQLITE_OPEN_CREATE | rusqlite::OpenFlags::SQLITE_OPEN_READ_WRITE,
+            false,
+        )?;
+
+        // check if the db needs to be instantiated regardless of whether or not
+        //  it was newly created: the db itself may be shared with other fee estimators,
+        //  which would not have created the necessary table for this estimator.
+        let tx = tx_begin_immediate_sqlite(&mut db)?;
+        Self::instantiate_db(&tx)?;
+        tx.commit()?;
+
+        Ok(Self {
+            db,
+            metric,
+            window_size,
+            full_block_weight: 6 * PROPORTION_RESOLUTION,
```

Comment:
Regarding my comment above about the block weight to use -- I think this should just be a single dimension fill (so just `PROPORTION_RESOLUTION`), because a block with that weight may not be able to add any more transactions.

Reply:
dropped the 6. some of the parameters in the tests changed, because the blocks got different amounts full.